semtypes/src/main/java/io/ballerina/types/Error.java

```diff
@@ -28,22 +37,24 @@
  */
 public class Error {
     public static SemType errorDetail(SemType detail) {
-        SubtypeData sd = Core.subtypeData(detail, BasicTypeCode.BT_MAPPING); // TODO: type should be MAPPING_RO
+        SubtypeData sd = bddIntersect((Bdd) subtypeData(detail, BasicTypeCode.BT_MAPPING), BDD_SUBTYPE_RO);
         if (sd instanceof AllOrNothingSubtype allOrNothingSubtype) {
             if (allOrNothingSubtype.isAllSubtype()) {
-                return PredefinedType.ERROR;
+                return ERROR;
             } else {
                 // XXX This should be reported as an error
-                return PredefinedType.NEVER;
+                return NEVER;
             }
-        } else {
-            return PredefinedType.basicSubtype(BasicTypeCode.BT_ERROR, (ProperSubtypeData) sd);
         }
+        if (sd == BDD_SUBTYPE_RO) {
+            return ERROR;
+        }
+        return basicSubtype(BT_ERROR, (ProperSubtypeData) sd);
     }
 
     // distinctId must be >= 0
     public SemType errorDistinct(int distinctId) {
-        BddNode bdd = BddCommonOps.bddAtom(RecAtom.createRecAtom(-distinctId - 1));
-        return PredefinedType.basicSubtype(BasicTypeCode.BT_ERROR, bdd);
+        BddNode bdd = bddAtom(createRecAtom(-distinctId - 1));
```

Comment:
Would it be better if we put an assertion here just in case, perhaps we can then get rid of the comment as well. WDYT?
```suggestion
        assert distinctId >= 0
        BddNode bdd = bddAtom(createRecAtom(-distinctId - 1));
```